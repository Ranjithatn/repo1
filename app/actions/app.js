import {
  UPDATE_SERVER_STATUS,
  REQUEST_NFIQ_QUALITY_THRESHOLD,
  REQUEST_NFIQ_QUALITY_THRESHOLD_SUCCESS,
} from "./actionTypes";

export const requestUpdateServerStatus = (data) => {
	return {
	  type: UPDATE_SERVER_STATUS,
	  payload: data
	};
}

export const requestNFIQQualityThreshold = (data) => {
	return {
	  type: REQUEST_NFIQ_QUALITY_THRESHOLD,
	  payload: data
	};
}

export const requestUpdateNFIQQualityThreshold = (data) => {
	return {
	  type: REQUEST_NFIQ_QUALITY_THRESHOLD_SUCCESS,
	  payload: data
	};
}

