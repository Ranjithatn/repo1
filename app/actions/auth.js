// @flow
import {
  REQUEST_LOGIN,
  REQUEST_LOGIN_SUCCESS,
  REQUEST_LOGIN_FAILED,
  REQUEST_LOGOUT,
  REQUEST_LOGOUT_SUCCESS,
  REQUEST_SET_NEW_UPDATE,
  REQUEST_SHOW_UPDATE_MESSAGE,
  REQUEST_SET_DOWNLOAD,
  REQUEST_SET_STATUS,
  REQUEST_CHECK_FOR_APP_UPDATE,
  REQUEST_LOGIN_POPUP,
  REQUEST_SET_LOOKUPS,
  REQUEST_SET_LOOKUP_CRIME_TYPES,
  REQUEST_SET_LOOKUP_NATIONALITY,
  REQUEST_SET_USER_INFO,
} from "./actionTypes";

type credentials = {
  username?: string,
  password?: string
};

export const requestLogin = (payload: credentials, history) => {
  return {
    type: REQUEST_LOGIN,
    payload,
    history
  };
};


export const requestLoginPopup = (payload: credentials) => {
  return {
    type: REQUEST_LOGIN_POPUP,
    payload,
  };
};


export const requestLoginSuccess = (authData: {
  jwt: string,
  roles: any,
  parsedRoles: array,
}) => {
  return {
    type: REQUEST_LOGIN_SUCCESS,
    payload: authData
  };
};

export const requestLoginFailed = (error: string) => {
  return {
    type: REQUEST_LOGIN_FAILED,
    payload: error
  };
};

export const requestLogout = history => {
  return {
    type: REQUEST_LOGOUT,
    history
  };
};

export const requestLogoutSuccess = () => {
  return {
    type: REQUEST_LOGOUT_SUCCESS
  };
};
export const requestSetNewUpdate = () => {
  return {
    type: REQUEST_SET_NEW_UPDATE
  };
};
export const requestSetDownload = version => {
  return {
    type: REQUEST_SET_DOWNLOAD,
    payload: version
  };
};
export const requestShowUpdateMessage = msg => {
  return {
    type: REQUEST_SHOW_UPDATE_MESSAGE,
    payload: msg
  };
};
export const requestSetStatus = status => {
  return {
    type: REQUEST_SET_STATUS,
    payload: status
  };
};



export const requestCheckForAppUpdate = () => {
  return {
    type: REQUEST_CHECK_FOR_APP_UPDATE,
  };
};

export const requestSetLookups = (data) => {
  return {
    type: REQUEST_SET_LOOKUPS,
    payload: data,
  }
}

export const requestSetLookupCrimeTypes = (data) => {
  return {
    type: REQUEST_SET_LOOKUP_CRIME_TYPES,
    payload: data,
  }
}

export const requestSetLookupNationality = (data) => {
  return {
    type: REQUEST_SET_LOOKUP_NATIONALITY,
    payload: data,
  }
}


export const requestSetUserInfo = (data) => {
  return {
    type: REQUEST_SET_USER_INFO,
    payload: data || {},
  }
}

