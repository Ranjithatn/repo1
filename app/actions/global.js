import {
  REQUEST_INPUT_FIELD_CHANGED,
  REQUEST_INPUT_FIELD_CHANGED_FAILED,
  RECEIVE_INPUT_FIELD_CHANGED,
  REQUEST_PAGE_CHANGED,
  REQUEST_PAGE_CHANGED_FAILED,
  REQUEST_TABLE_ROW_CLICKED,
  REQUEST_TABLE_ROW_CLICKED_FAILED,
  RECEIVE_TABLE_ROW_CLICKED
} from "./actionTypes";

export const requestInputFieldChanged = (e, area: string) => {
  const { id, value } = e.target;
  const data = {
    field: id,
    value: value,
    area: area
  };

  return {
    type: REQUEST_INPUT_FIELD_CHANGED,
    payload: data
  };
};

type FieldData = {
  field: ?string,
  value: string,
  area: string
};

type TableRowData = {
  area: string,
  dataId: ?string
};

export const receiveInputFieldChanged = (data: FieldData) => {
  return {
    type: RECEIVE_INPUT_FIELD_CHANGED,
    payload: data
  };
};

export const receiveInputFieldChangedFailed = () => ({
  type: REQUEST_INPUT_FIELD_CHANGED_FAILED,
});

// e, area: string
export const requestPageChanged = (e, area) => {

  if (!e.target || !e.target.tagName || e.target.tagName !== "A")
    //only send proper action if link is clicked
    return { type: "unknown", payload: {} };
  const pageNumberClicked = e.target.id;

  return {
    type: REQUEST_PAGE_CHANGED,
    payload: {
      page: pageNumberClicked,
      area
    }
  };
};

export const requestPageChangedFailed = () => ({
  type: REQUEST_PAGE_CHANGED_FAILED
});

export const requestTableRowClicked = (dataId, area, status, type,tcn) => {
  const data = {
    dataId,
    area,
    status,
    type,
    tcn
  };
  return {
    type: REQUEST_TABLE_ROW_CLICKED,
    payload: data
  };
};

export const requestTableRowClickedFailed = (dataId, area: string) => ({
  type: REQUEST_TABLE_ROW_CLICKED_FAILED
});

export const receiveTableRowClicked = (payload: TableRowData) => {
  return {
    type: RECEIVE_TABLE_ROW_CLICKED,
    payload
  };
};
