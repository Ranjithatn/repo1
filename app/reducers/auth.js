//@flow
import {
  REQUEST_LOGIN,
  REQUEST_LOGIN_SUCCESS,
  REQUEST_LOGIN_FAILED,
  REQUEST_LOGOUT_SUCCESS,
  RECEIVE_INPUT_FIELD_CHANGED,
  REQUEST_SET_NEW_UPDATE,
  REQUEST_SHOW_UPDATE_MESSAGE,
  REQUEST_SET_DOWNLOAD,
  REQUEST_SET_STATUS,
  REQUEST_SET_LOOKUPS,
  REQUEST_SAVE_ANNOTATIONS_CUSTOM,
  REQUEST_SYSTEM_PERMISSION_SUCCESS,
  REQUEST_SET_LOOKUP_CRIME_TYPES,
  REQUEST_SET_LOOKUP_NATIONALITY,
  REQUEST_SET_USER_INFO,
} from "../actions/actionTypes";

const initialState = {
  username: "",
  password: "",
  apiUri: "",
  jwt: "",
  error: "",
  roles: "",
  newUpdate: undefined,
  updateMessage: "",
  newDownload: undefined,
  version: "",
  updateStatus: "",
  parsedRoles: [],
  lookups: [],
  menuPerm:{},
  servicePerm:{}
};

export default function login(state = initialState, action = {}) {
  switch (action.type) {
    case REQUEST_LOGIN:
      return { ...state, error: "" };
    case REQUEST_LOGIN_SUCCESS:
      const { jwt, roles, parsedRoles } = action.payload;
      return { ...state, password: "", jwt, roles, parsedRoles };
    case REQUEST_LOGIN_FAILED:
      return { ...initialState, error: action.payload };
    case REQUEST_LOGOUT_SUCCESS:
      localStorage.removeItem("jwt");
      return initialState;

      return initialState;
    case RECEIVE_INPUT_FIELD_CHANGED:
      if (action.payload.area === "auth") {
        const { field, value } = action.payload;
        let newState = Object.assign({}, state);
        newState[field] = value;
        return newState;
      } else {
        return state;
      }
    case REQUEST_SET_NEW_UPDATE:
      return { ...state, newUpdate: true };
    case REQUEST_SET_STATUS:
      return { ...state, updateStatus: action.payload };
    case REQUEST_SET_DOWNLOAD:
      return { ...state, newDownload: true, version: action.payload };
    case REQUEST_SHOW_UPDATE_MESSAGE:
      return { ...state, updateMessage: action.payload };
    case REQUEST_SET_LOOKUPS:
      return { ...state, lookups: action.payload };
    case REQUEST_SYSTEM_PERMISSION_SUCCESS:
      return {
        ...state,
        // servicePerm: action.payload.service_perm,
        menuPerm: action.payload.menu_perm
      };
    case REQUEST_SET_LOOKUP_CRIME_TYPES:
      return { ...state, lookupCrimeTypes: action.payload };

    case REQUEST_SET_LOOKUP_NATIONALITY:
      return { ...state, nationality: action.payload };

    case REQUEST_SET_USER_INFO:
     return { ...state, userInfo: action.payload };



    default:
      return state;
  }
}
