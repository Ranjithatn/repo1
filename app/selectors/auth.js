import { createSelector } from "reselect";

export const usernameSelector = createSelector(
  state => state.auth.username,
  u => u
);
export const updateMessageSelector = createSelector(
  state => state.auth.updateMessage,
  updateMsg => updateMsg
);
export const downloadSelector = createSelector(
  state => state.auth.newDownload,
  d => d
);
export const updateMsgSelector = createSelector(
  state => state.auth.updateStatus,
  updateStatus => updateStatus
);

export const lookupsSelector = createSelector(
  state => state.auth.lookups,
  lookups => lookups
);



export const userLocationSelector = createSelector(
  state => state.auth.lookups,
  (lookups) => {
    console.log("userLocationSelector::lookups", lookups);
    let result = [];

    if ( lookups.length > 0 ) {
      lookups.find( item => {
        if ( item.lookupName === "LocationIDs" ) {
          result = item.items;
          return true;
        }
      });
    }

    console.log("userLocationSelector::result",result);

    return result;
  }
);

export const lookupCrimeTypesSelector = createSelector(
  state => state.auth.lookupCrimeTypes,
  lookupCrimeTypes => lookupCrimeTypes
);

