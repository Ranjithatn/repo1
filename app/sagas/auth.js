import { call, put, takeLatest, select } from "redux-saga/effects";
import Constants from "../constants";
import * as localStorage from "../utils/localStorage";
import { delay } from "redux-saga";
import { push } from "react-router-redux";
// import { autoUpdater } from "electron-updater";
import {
  REQUEST_LOGIN,
  REQUEST_LOGIN_SUCCESS,
  REQUEST_LOGOUT,
  REQUEST_LOGOUT_SUCCESS,
  REQUEST_SPINNER_START,
  REQUEST_SPINNER_STOP,
  REQUEST_CHECK_FOR_APP_UPDATE,
  REQUEST_LOGIN_POPUP
} from "../actions/actionTypes";
import { requestInputFieldChanged } from '../actions/global';
import {
  requestLoginSuccess,
  requestLogoutSuccess,
  requestLoginFailed,
  requestSetLookups,
  requestSetLookupCrimeTypes,
  requestSetLookupNationality,
  requestSetUserInfo,
} from "../actions/auth";
import { requestShowNotification } from "../actions/notifications";
import {
  // updateServerConnectionStatus,
  requestUpdateServerStatus
} from "../actions/app";
import { updateMsgSelector } from "../selectors/auth";
import { requestActiveJobs, requestSystemPermissions } from "../actions/jobs";
import Api, { url } from "../api/";
import { startSpinner, stopSpinner } from "../../app/actions/spinner";
import { settings } from "../utils/electron";

import { store } from "../index";
import path from "path";
import { translateRes } from "../utils/intl";
import { displayLoginPopup } from "../components/Notification/PortalNotification";

import moment from "moment";


const LivescanAPI = require("../hardwareSDK/biocoreSdk/livescan/livescan");

const electron = require("electron").remote;

const ipcRenderer = require("electron").ipcRenderer;

export function* watchRequestLogin() {
  yield takeLatest(REQUEST_LOGIN, callLogin);
}

export function* watchRequestLoginPopup() {
  yield takeLatest(REQUEST_LOGIN_POPUP, callLoginPopup);
}

export function* watchCheckForappUpdate() {
  yield takeLatest(REQUEST_CHECK_FOR_APP_UPDATE, checkForAppUpdate);
}

export function* watchRequestLogout() {
  yield takeLatest(REQUEST_LOGOUT, callLogout);
}

export function* checkForAppUpdate(action) {
  const _settings = settings();

  let newApiUri = "";
  if (_settings && _settings.ipAddress) {
    newApiUri = `${_settings.protocol ? _settings.protocol : "http"}://${
      _settings.ipAddress
    }/services-gateway-ucw/api`;
    Constants.API_ORIGIN_URL = newApiUri;
    localStorage.set("api", newApiUri);
  }

  yield put(startSpinner());
  yield call(delay, 1000);
  ipcRenderer.send("checkForUpdates");
}

export function* callLogin(action) {
  yield put(stopSpinner());
  try {
    yield put(startSpinner());
    const { payload } = action;
    const _settings = settings();

    let newApiUri = "";
    if (_settings && _settings.ipAddress) {
      newApiUri = `${_settings.protocol ? _settings.protocol : "http"}://${
        _settings.ipAddress
      }/services-gateway-ucw/api`;
      Constants.API_ORIGIN_URL = newApiUri;
      localStorage.set("api", newApiUri);
    }
    // localStorage.set("api", Constants.API_ORIGIN_URL);

    const loginRes = yield call(Api, {
      url: url.auth.login,
      method: "POST",
      data: payload
    });
    if (loginRes.jwt) {
      const { jwt } = loginRes;
      // const {
      //   "http://schemas.microsoft.com/ws/2008/06/identity/claims/role": roles
      // } = yield call(parseJwt, jwt);

      const parsed = yield call(parseJwt, jwt);

      // console.log( "userRoles", parsed.userRoles );
      // console.log( "userRoles::JSON", JSON.parse(parsed.userRoles) );
      // const parsedRoles = JSON.parse(parsed.userRoles);
      const parsedRoles = parsed.userRoles;

      let {
        "http://schemas.microsoft.com/ws/2008/06/identity/claims/role": roles
      } = parsed;

      // console.log("AUTH::roles",roles);
      if (!roles) {
        roles = "custom";
      }

      yield call(localStorage.set, "jwt", jwt);

      yield put(requestLoginSuccess({ jwt, roles, parsedRoles }));
      yield put( requestSystemPermissions() );
      yield call( requestLookupCrimeTypes );
      yield call( getUserLocations );
      // yield call( requestLookupNationality );

      yield put( requestSetUserInfo(loginRes.userInfo) );

      yield put(requestActiveJobs());
      yield put(push("/authenticated/jobqueue"));

      yield put(stopSpinner());

      if (localStorage.get("loginpopup")) {
        localStorage.remove("loginpopup");
      }

      yield call(startCheckingServerConnectionStatus);
      yield call(checkForTokenExpiration);
    } else {
      if (loginRes.Error) {
        yield put(requestLoginFailed(loginRes.Error));
      }
      yield put(stopSpinner());
      yield put(requestInputFieldChanged( { target: { id: 'username', value: '', } }, 'auth'));
      yield put(requestInputFieldChanged( { target: { id: 'password', value: '', } }, 'auth'));
    }
  } catch (e) {
    yield put(requestLoginFailed("Login failed for reasons unknown"));
  } finally {
    yield put(stopSpinner());
    // yield call(startCheckingServerConnectionStatus);
  }
}

export function* callLoginPopup(action) {
  try {
    localStorage.set("loginpopup", "true");
    yield put(startSpinner());
    const { payload } = action;
    const _settings = settings();

    let newApiUri = "";
    if (_settings && _settings.ipAddress) {
      newApiUri = `${_settings.protocol ? _settings.protocol : "http"}://${
        _settings.ipAddress
      }/services-gateway-ucw/api`;
      Constants.API_ORIGIN_URL = newApiUri;
      localStorage.set("api", newApiUri);
    }
    localStorage.set("api", Constants.API_ORIGIN_URL);

    const loginRes = yield call(Api, {
      url: url.auth.login,
      method: "POST",
      data: payload
    });

    if (loginRes.jwt) {
      const { jwt } = loginRes;

      const parsed = yield call(parseJwt, jwt);

      // const parsedRoles = JSON.parse(parsed.userRoles);
      const parsedRoles = parsed.userRoles;

      let {
        "http://schemas.microsoft.com/ws/2008/06/identity/claims/role": roles
      } = parsed;

      if (!roles) {
        roles = "custom";
      }

      yield call(localStorage.set, "jwt", jwt);

      yield put(requestLoginSuccess({ jwt, roles, parsedRoles }));
      yield put(requestSystemPermissions())

      yield call( requestLookupCrimeTypes );
      yield call( getUserLocations );
      // yield call( requestLookupNationality );

      yield put( requestSetUserInfo(loginRes.userInfo) );

      yield put(requestActiveJobs());
      // yield put(push("/authenticated/jobqueue"));
      localStorage.remove("loginpopup");

      yield put(stopSpinner());
    } else {
      if (loginRes.props.message) {
        // displayLoginPopup( loginRes.props.message );
      }
      yield put(stopSpinner());
    }
  } catch (e) {
    displayLoginPopup("Login failed for reasons unknown");
  } finally {
    yield put(stopSpinner());
  }
}

export function* callLogout(action) {
  yield put(startSpinner());
  try {
    const { history } = action;
    const logoutRes = yield call(Api, {
      url: url.auth.logout,
      method: "POST",
      isJwtRequired: true,
      data: {}
    });
    if (logoutRes) {
      if (logoutRes.message) {
        yield put(
          requestShowNotification({
            message: translateRes(logoutRes.message),
            type: "is-success"
          })
        );
      }

      // stop livescan
      LivescanAPI.stopCapture();
      // stop cardscan
      require("dwt");

      Dynamsoft.WebTwainEnv.Unload();

      yield call(localStorage.set, "jwt", "");
      yield put(requestLogoutSuccess());
      yield put(push("/"));
    }
  } catch (e) {
    console.error("-------- logout failed", e);
  } finally {
    yield put(stopSpinner());
  }
}

export function parseJwt(token) {
  var base64Url = token && token.split(".")[1];
  var base64 = base64Url.replace("-", "+").replace("_", "/");

  const parsed = JSON.parse(window.atob(base64));

  return parsed;
}

export function* handleConnectionStatus() {
  const connStatus = yield call(Api, {
    url: url.systeminfo.systeminfo,
    method: "GET",
    isJwtRequired: true,
    data: {}
  });

  if (connStatus) {
    // console.log("connectionStatus API response", connStatus);
    if (connStatus.serverLiveStatus && connStatus.serverLiveStatus.length > 0) {
      let status = "unknown";
      if (connStatus.serverLiveStatus[0].isActive) {
        status = "connected";
      } else {
        status = "disconnected";
      }
      yield put(requestUpdateServerStatus(status));
    }
  } // if ends here
}

export function* startCheckingServerConnectionStatus() {
  const _settings = settings();
  let callTime = 0;

  if (_settings.serverConnection && _settings.serverConnection > 0) {
    callTime = _settings.serverConnection;
  }

  if (callTime > 0) {
    while (localStorage.get("jwt") && localStorage.get("jwt") != "") {
      // console.log("inside while loop");
      // user is logged in lets continue
      yield call(handleConnectionStatus);
      yield call(delay, callTime * 1000);
    }
  } else {
    // console.log("server connection time is not specified");
    yield call(handleConnectionStatus);
  }
}

export function* getUserLocations() {
  // console.log("getUserLocations::init");

  try {
    const apiCall = yield call(Api, {
      url: url.systeminfo.lookups,
      method: "GET",
      isJwtRequired: true,
      data: {}
    });

    if (apiCall) {
      // console.log("apiCall::success", apiCall);
      yield put(requestSetLookups(apiCall));
    } else {
      console.log("apiCall::failed", apiCall);
    }
  } catch (e) {
    console.log("getUserLocations::catch::error occoured", e);
  }
}


export function* requestLookupCrimeTypes() {
  try {
    const apiCall = yield call(Api, {
      url: url.systeminfo.getniclookupcrimetypes,
      method: "GET",
      isJwtRequired: true,
      data: {}
    });

    if (apiCall) {
      // console.log("requestLookupCrimeTypes::apiCall::success", apiCall);
      yield put( requestSetLookupCrimeTypes(apiCall) );
    } else {
      // console.log("requestLookupCrimeTypes::apiCall::failed", apiCall);
    }
  } catch (e) {
    console.log("requestLookupCrimeTypes::catch::error occoured", e);
  }

}




export function* requestLookupNationality() {
  try {
    const apiCall = yield call(Api, {
      url: url.systeminfo.nationality,
      method: "GET",
      isJwtRequired: true,
      data: {},
      fromBase: true,
    });

    if (apiCall) {
      yield put( requestSetLookupNationality(apiCall) );
    }
  } catch (e) {
    console.log("requestLookupNationality::catch::error occoured", e);
  }

}






export function* handleTokenExpiration() {
  console.log("handleTokenExpiration");

  try {

    const jwt = localStorage.get("jwt");

    if ( jwt ) {
      const parsed = parseJwt(jwt);
      const expiryDate = new Date( parsed.exp * 1000 );
      const currentDate = new Date();

      let a = moment(expiryDate);
      let b = moment(currentDate);
      let minsLeft = a.diff(b, 'minutes');
      console.log("minutes left until token expiry",minsLeft);


      if ( minsLeft <= 1 ) {
            
        const newToken = yield call(Api, {
          url: url.auth.refreshToken,
          method: "POST",
          isJwtRequired: true,
          data: {}
        });

        console.log("newToken",newToken);
        if ( newToken ) {
          // localStorage.set("jwt", newToken);
          yield call(localStorage.set, "jwt", newToken.jwt);
        }

      }

    }

  } catch (e) {
    console.log("handleTokenExpiration::catch",e);
  }


}



export function* checkForTokenExpiration() {
  let callTime = 60000; // in ms

  while (localStorage.get("jwt") && localStorage.get("jwt") != "") {
    yield call(handleTokenExpiration);
    yield call(delay, callTime);
  }

}