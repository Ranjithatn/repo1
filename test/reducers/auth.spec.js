import auth from "../../app/reducers/auth";
import { spy } from "sinon";
import {
  requestLogin,
  requestLoginSuccess,
  requestLogoutSuccess,
  requestLoginFailed,
  requestSetNewUpdate,
  requestSetStatus,
  requestSetDownload,
  requestShowUpdateMessage
} from "../../app/actions/auth";
import { receiveInputFieldChanged } from "../../app/actions/global";

describe("auth reducers", () => {
  const initialState = {
    username: "",
    password: "",
    apiUri: "",
    jwt: "",
    error: "",
    roles: "",
    newUpdate: undefined,
    updateMessage: "",
    newDownload: undefined,
    version: "",
    updateStatus: "",
    parsedRoles: undefined,
    lookups: [],
  };
  xit("should have initial state", () => {
    const output = auth();
    expect(output).toEqual({...initialState, parsedRoles: []});
  });
  it("should handle initial state", () => {
    expect(auth(undefined, {})).toMatchSnapshot();
  });

  it("should handle RECEIVE_INPUT_FIELD_CHANGED", () => {
    const action = receiveInputFieldChanged({
      field: "username",
      value: "user",
      area: "auth"
    });
    const output = auth(
      { username: "t", password: "t", jwt: "", error: "" },
      action
    );
    const expected = {
      username: "user",
      password: "t",
      jwt: "",
      error: ""
    };
    expect(output).toEqual(expected);
  });

  it("should handle RECEIVE_INPUT_FIELD_CHANGED by returning prev state", () => {
    const action = receiveInputFieldChanged({
      field: "username",
      value: "user",
      area: "anotherArea"
    });
    const output = auth(initialState, action);
    expect(output).toEqual(initialState);
  });
  it("should clear old error message when user tries to login", () => {
    const historyMock = {
      push: spy()
    };
    const action = requestLogin(
      { username: "abcd", password: "abcde" },
      historyMock
    );
    const output = auth(
      {
        username: "abc",
        password: "abcd",
        error: "Invalid username or password"
      },
      action
    );
    const expected = {
      username: "abc",
      password: "abcd",
      error: ""
    };
    expect(output).toEqual(expected);
  });

  xit("should handle login success", () => {
    const initialState = {
      username: "t",
      password: "abc",
      roles: "",
      error: "",
      jwt: "",
      parsedRoles: undefined,
    };
    const action = requestLoginSuccess("abc");
    const output = auth(initialState, action);
    const expected = {
      username: "t",
      jwt: "abc",
      password: "",
      roles: "",
      error: ""
    };
    expect(output).toEqual(expected);
  });

  xit("should handle login failed", () => {
    const initialState = {
      username: "t",
      password: "abc",
      error: "",
      jwt: "",
      apiUri: ""
    };
    const action = requestLoginFailed("Invalid username or password");
    const output = auth(initialState, action);
    const expected = {
      username: "",
      jwt: "",
      password: "",
      roles: "",
      apiUri: "",
      error: "Invalid username or password",
      newUpdate: undefined,
      updateMessage: "",
      newDownload: undefined,
      version: "",
      updateStatus: "",
      parsedRoles: [],
      lookups: [],
    };
    expect(output).toEqual(expected);
  });
  it("should handle login success", () => {
    const action = requestLoginSuccess({ jwt: "jwt", roles: "admin" });
    const output = auth(initialState, action);
    const expected = {
      ...initialState,
      jwt: "jwt",
      roles: "admin"
    };
    expect(output).toEqual(expected);
  });

  xit("should handle logout success", () => {
    const initialState = {
      error: "",
      username: "a",
      password: "",
      jwt: "abc"
    };
    const action = requestLogoutSuccess();
    const output = auth(...initialState, action);
    const expected = {
      ...initialState,
      apiUri: "",
      error: "",
      jwt: "jwt",
      newDownload: undefined,
      newUpdate: undefined,
      password: "",
      roles: "admin",
      updateMessage: "",
      updateStatus: "",
      username: "",
      version: "",
      parsedRoles: [],
    };
    expect(output).toEqual(expected);
  });

  it("should handle unknown action type", () => {
    const initialState = {
      username: "a",
      password: "",
      jwt: "abc",
      error: ""
    };
    const action = {
      type: "unknown"
    };
    const output = auth(initialState, action);
    const expected = { username: "a", password: "", jwt: "abc", error: "" };
    expect(output).toEqual(expected);
  });
  it("should handle REQUEST_SET_NEW_UPDATE type", () => {
    const action = requestSetNewUpdate();
    const output = auth(initialState, action);
    const expected = { ...initialState, newUpdate: true };
    expect(output).toEqual(expected);
  });
  it("should handle REQUEST_SET_STATUS type", () => {
    const action = requestSetStatus("data");
    const output = auth(initialState, action);
    const expected = { ...initialState, updateStatus: "data" };
    expect(output).toEqual(expected);
  });
  it("should handle REQUEST_SET_DOWNLOAD type", () => {
    const action = requestSetDownload(1);
    const output = auth(initialState, action);
    const expected = { ...initialState, newDownload: true, version: 1 };
    expect(output).toEqual(expected);
  });
  it("should handle REQUEST_SHOW_UPDATE_MESSAGE type", () => {
    const action = requestShowUpdateMessage("data");
    const output = auth(initialState, action);
    const expected = { ...initialState,updateMessage:"data" };
    expect(output).toEqual(expected);
  });
});
