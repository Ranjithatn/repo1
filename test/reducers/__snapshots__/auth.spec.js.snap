// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth reducers should handle initial state 1`] = `
Object {
  "apiUri": "",
  "error": "",
  "jwt": "",
  "lookups": Array [],
  "menuPerm": Object {},
  "newDownload": undefined,
  "newUpdate": undefined,
  "parsedRoles": Array [],
  "password": "",
  "roles": "",
  "servicePerm": Object {},
  "updateMessage": "",
  "updateStatus": "",
  "username": "",
  "version": "",
}
`;
